src/transforms/geoip.rs

```diff
@@ -0,0 +1,116 @@
+extern crate maxminddb;
+
+use super::Transform;
+
+use crate::{
+    event::Event,
+    topology::config::{DataType, TransformConfig},
+};
+use serde::{Deserialize, Serialize};
+use string_cache::DefaultAtom as Atom;
+
+use std::net::IpAddr;
+use std::str::FromStr;
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct GeoipConfig {
+    pub source: Atom,
+    pub database: String,
+    pub target: String,
+}
+
+pub struct Geoip {
+    pub source: Atom,
+    pub database: String,
+    pub target: String,
+}
+
+#[typetag::serde(name = "geoip")]
+impl TransformConfig for GeoipConfig {
+    fn build(&self) -> Result<Box<dyn Transform>, crate::Error> {
+        Ok(Box::new(Geoip::new(
+            self.source.clone(),
+            self.database.clone(),
+            self.target.clone(),
+        )))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+impl Geoip {
+    pub fn new(source: Atom, database: String, target: String) -> Self {
+        Geoip {
+            source: source,
+            database: database,
+            target: target,
+        }
+    }
+}
+
+impl Transform for Geoip {
+    fn transform(&mut self, mut event: Event) -> Option<Event> {
+        println!("Event: {:?}", event.as_log());
+        let ipaddress = event
+            .as_log()
+            .get(&self.source)
+            .map(|s| s.to_string_lossy());
+        if let Some(ipaddress) = &ipaddress {
+            let ip: IpAddr = FromStr::from_str(ipaddress).unwrap();
+            println!("Looking up {}", ip);
+            let reader =
+                maxminddb::Reader::open_readfile("/usr/local/share/GeoIP/GeoIP2-City.mmdb")
+                    .unwrap();
+            let city: maxminddb::geoip2::City = reader.lookup(ip).unwrap();
```

Comment:
We'll need to handle the case where this lookup fails. I think a sensible default behavior would just be to do nothing and pass through the event as it was.

Reply:
I agree. How does the updated code look?